src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol.java

```diff
@@ -96,9 +97,12 @@ abstract boolean performHandshake(DataOutputStream outputStream,
      *
      * @param socket The established {@link Socket} connection to the master.
      * @param channelBuilder The {@link ChannelBuilder} to use.
+     * @param inputStream The {@link BufferedInputStream} of {@link Socket#getInputStream()} or {@code null} if there
+     *                    is none (TODO remove the null possibility).
      * @return The constructed channel
      */
-    abstract Channel buildChannel(Socket socket, ChannelBuilder channelBuilder) throws IOException;
+    abstract Channel buildChannel(Socket socket, ChannelBuilder channelBuilder,
+                                  BufferedInputStream inputStream) throws IOException;
```

Comment:
Perhaps it would make more sense to change the method interface to:

abstract Channel buildChannel(ChannelBuilder channelBuilder, BufferedInputStream inputStream, BufferedOutputStream outputStream) throws IOException;

Or just use InputStream and OutputStream in the method signature. That way implementations will not make the mistake of trying to get the inputstream directly from the socket.